polonius-parser/src/lexer.rs

```diff
@@ -0,0 +1,185 @@
+                    .unwrap_or(0) as u32
+                    + 1,
+                match c {
+                    b'\'' => T![origin],
+                    b'B' => T![Block],
+                    b'L' => T![loan],
+                    b'V' => T![variable],
+                    _ => unreachable!(),
+                },
+            ),
+            [b'/', b'/', ..] => (
+                input
+                    .char_indices()
+                    .take_while(|(_, c)| *c != '\n')
+                    .last()
+                    .map(|(pos, _c)| pos)
+                    .unwrap_or(input.len()) as u32
+                    + 1,
+                T![comment],
+            ),
+            // relation keywords
+            kw if kw.starts_with("use_of_var_derefs_origin".as_bytes()) => (
+                "use_of_var_derefs_origin".len() as u32,
+                T![use_of_var_derefs_origin],
+            ),
+            kw if kw.starts_with("drop_of_var_derefs_origin".as_bytes()) => (
+                "drop_of_var_derefs_origin".len() as u32,
+                T![drop_of_var_derefs_origin],
+            ),
+            kw if kw.starts_with("placeholders".as_bytes()) => {
+                ("placeholders".len() as u32, T![placeholders])
+            }
+            kw if kw.starts_with("known_subsets".as_bytes()) => {
+                ("known_subsets".len() as u32, T![known subsets])
+            }
+            // CFG keywords
+            kw if kw.starts_with("block".as_bytes()) => ("block".len() as u32, T![block]),
+            kw if kw.starts_with("goto".as_bytes()) => ("goto".len() as u32, T![goto]),
+            // effect keywords - facts
+            kw if kw.starts_with("outlives".as_bytes()) => ("outlives".len() as u32, T![outlives]),
+            kw if kw.starts_with("loan_issued_at".as_bytes()) => {
+                ("loan_issued_at".len() as u32, T![loan_issued_at])
+            }
+            kw if kw.starts_with("loan_invalidated_at".as_bytes()) => {
+                ("loan_invalidated_at".len() as u32, T![loan_invalidated_at])
+            }
+            kw if kw.starts_with("loan_killed_at".as_bytes()) => {
+                ("loan_killed_at".len() as u32, T![loan_killed_at])
+            }
+            kw if kw.starts_with("var_used_at".as_bytes()) => {
+                ("var_used_at".len() as u32, T![var_used_at])
+            }
+            kw if kw.starts_with("var_defined_at".as_bytes()) => {
+                ("var_defined_at".len() as u32, T![var_defined_at])
+            }
+            kw if kw.starts_with("origin_live_on_entry".as_bytes()) => (
+                "origin_live_on_entry".len() as u32,
+                T![origin_live_on_entry],
+            ),
+            kw if kw.starts_with("var_dropped_at".as_bytes()) => {
+                ("var_dropped_at".len() as u32, T![var_dropped_at])
+            }
+            // effect keywords - use
+            kw if kw.starts_with("use".as_bytes()) => ("use".len() as u32, T![use]),
```

Comment:
Why prefix-like tree structure ([trie](https://en.wikipedia.org/wiki/Trie)) is not used here? It seems like it can provide additional performance because it considers all keywords simultaneously during comparison.

Reply:
Mostly because:
- the goal was to reduce compile times, and runtime is not especially important here: the tests are small, and not numerous yet. The slight gain in parsing could be interesting in the future, but not essential for this to land. If you're interested in benchmarking and improving this, by all means please do, we would love that.
- this whole PR was done during the latest friday-afternoon-sprint, and the lexer in particular was put together in a couple hours at most. Impressive work in such a short time.